### PR TITLE
[DO NOT MERGE] Android Virtual Assistant Sample App

### PR DESCRIPTION
**Pull request by dwhiteddsoft**
This is the first PR for issue #331 . There will be work past this as discussed. 
To use, from the menu configure your background color, bot secret, speech recognition key and if you have one a speech endpoint. 
Press the start conversation button and it will interact with the bot. Putting configurable user into the app is the next pull request and a wider testing of bots. 
Adaptive cards will render how ever they are configured from the server using the AdaptiveCard library.   Returns with no adaptive card and simply text, will render simply as text. If you have a card with speak filled in, it will speak that as well.

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            _chatbot.send("", "endOfConversation", "", _channleData);
+        }
+    }
+
+    //returns the conversationID
+    private void startConversation() {
+        Log.d(TAG,"Bot Secret: " + primaryToken);
+        _chatbot = new DirectLineChatbot(primaryToken);
+        _chatbot.setDebug(true);
+        _chatbot.setUser("MeganB@GMIPATest.OnMicrosoft.com");
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:50]: User will actually be a UserID (GUID) and should be configurable

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            @Override
+            public void onMessageReceived(@NotNull String message)
+            {
+                try {
+                    JSONObject json = new JSONObject(message);
+                    String errorText = json.optString("text",null);
+                    if (errorText != null)
+                        _botEventInterface.onBotError(errorText);
+                    else {
+                        //Log.d("CHATBOT RECV: ", message);
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:72]: Can we do a pass and remove older code?

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                }
+            }
+        });
+        _botEventInterface.onBotReady();
+    }
+
+    private void PopulateUserInfo() {
+        GeoLocation geo = new GeoLocation(47.7069312,-122.09520640000001);
+        Claims claims = new Claims("true");
+        _channleData = new ChannelData("MeganVIN101",
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:88]: No need to send ChannelData in this way for normal bots, just the older IPA stuff.  Location is needed, can we instead get this from the device and fall-back to a (configurable) default? Thanks

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+    };
+    Handler processHandler = new Handler();
+    Runnable processRunnable = new Runnable() {
+        @Override
+        public void run() {
+            if (!_bRecognizing) {
+                if (!processingStack.isEmpty()) {
+                    if (_bIsSpeechReady) {
+                        String msg = processingStack.remove(0);
+                        //Log.d(TAG, "Processing msg: " + msg);
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java:73]: Can we add the logging back in (we use this approach in the Speech Devices SDK), will just help diagnose issues

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+                _bot.BotState = _bot.BOT_STATE_ACCEPTING_INPUT;
+            else if (input.contains("ignoringInput"))
+                _bot.BotState = _bot.BOT_STATE_IGNORING_INPUT;
+            else if (input.length() == 0)
+                _bot.BotState = _bot.BOT_STATE_ACCEPTING_INPUT;
+            else
+                Log.d(TAG,"input hint is not what we expect: " + input);
+
+            if (type.compareTo("message") == 0) {
+                //((TextView) (getActivity().findViewById(R.id.txtSpeechDisplay))).setText("");
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java:194]: Remove this older code?

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+                    layout.removeAllViewsInLayout();
+                    TextView lblTV = new TextView(this);
+                    lblTV.setText("Returned text: " + text);
+                    lblTV.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
+                    lblTV.setTextColor(Color.parseColor("#000000"));
+                    lblTV.setTextSize(16f);
+                    lblTV.setGravity(Gravity.CENTER);
+                    layout.addView(lblTV);
+                }
+                if (speak.length() > 0) {
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java:223]: Might want to be defensive, in cases where .Speak is not set but .Text is set fall-back to this (and perhaps log that a Speak variant was not provided)

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+    }
+    //ICardActionHandler implementation
+
+    //speech interface
+    @Override
+    public void onSpeechRecognizing(String text) {
+        runOnUiThread(new Runnable() {
+            public void run() {
+                try {
+                    //((TextView)(getActivity().findViewById(R.id.txtSpeechDisplay))).setText(text);
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/MainActivity.java:450]: ?

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/speech/SpeechImpl.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+                    if (langResult == TextToSpeech.LANG_MISSING_DATA || langResult == TextToSpeech.LANG_NOT_SUPPORTED) {
+                        tts.setLanguage(Locale.US);
+                    }
+                    _speechEventInterface.onInitialized();
+                }
+            }
+        });
+        tts.setOnUtteranceProgressListener(mProgressListener);
+
+//        try {
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/speech/SpeechImpl.java:86]: remove?

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/DirectLineChatbot.kt [hunk 9]
```diff
@@ -0,0 +1,204 @@
+// The MIT License (MIT)
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/DirectLineChatbot.kt:1]: Is this a library from a 3rd party? If so, can you point me at the original source so I can get the use validated before we merge please?

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
+package com.microsoft.assistant_android;
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:1]: In this and all other source files can we add this copyright header please?

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/readme.md [hunk 11]
```diff
@@ -0,0 +1,4 @@
+# DirectLineChatBot code
+This is borrowed and modified from https://github.com/smartnsoft/directline-chatbot-android-sdk
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/readme.md:2]: Got it (see other comment). I've asked for a check

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/DirectLineChatbot.kt (hunk 9 above)
Comment (darrenj) [reply solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/DirectLineChatbot.kt:1]: Looks like @ryanlengel didn't have to use this library and still made use of websockets (see [here](https://github.com/Microsoft/AI/blob/android-assistant/solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-SpeechDevicesSDK/virtualassistant/app/src/main/java/com/microsoft/bot/solutions/speechdevices/samples/botapp/MainActivity.java)). @ryanlengel can you confirm?

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+import com.microsoft.directlinechatbot.bo.GeoLocation;
+
+public class BotWrapper {
+    private static String TAG = "Bot";
+    private String primaryToken;
+    private DirectLineChatbot _chatbot = null;
+    private ChannelData _channleData = null;
+    private BotEventInterface _botEventInterface;
+
+    public String botName = "gmipapoc-botchannel";
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:21]: customer specific, needs removing

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    public void EndConversation() {
+        if (_chatbot != null) {
+            _chatbot.send("", "endOfConversation", "", _channleData);
+        }
+    }
+
+    //returns the conversationID
+    private void startConversation() {
+        Log.d(TAG,"Bot Secret: " + primaryToken);
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:47]: shouldn't log token

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java (hunk 1 above)
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:50]: should be configurable (and will be a guid)

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
+
+        StrictMode.setThreadPolicy(policy);
+        _chatbot.start(new DirectLineChatbot.Callback()
+        {
+            @Override
+            public void onStarted()
+            {
+                _chatbot.send("OpenAssistant","event", "", _channleData);
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:60]: This isn't an event needed - the demo bot used this approach but not the way forward. See other comment - we will send an event (IPA.Location) at startup or when location changes.

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+                    _botEventInterface.onBotError("Error in BotWrapper::onMessageReceived. Error: " + e.getMessage());
+                }
+            }
+        });
+        _botEventInterface.onBotReady();
+    }
+
+    private void PopulateUserInfo() {
+        GeoLocation geo = new GeoLocation(47.7069312,-122.09520640000001);
+        Claims claims = new Claims("true");
```
Comment (darrenj) [inline solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:87]: claims not needed

**Review by darrenj — CHANGES_REQUESTED** [review]
Thanks David. Some comments below that we should address ahead of merging. Also asking @ryanlengel to do a quick pass as he wrote the Speech DDK app and is closer to android dev than me.

A few other general things:

- [ ] Add README.MD in root directory with getting started instructions (as per [this](https://github.com/Microsoft/AI/tree/96c5042017f67f9e64ad9801768dc5603b721b5d/solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-SpeechDevicesSDK) example). Keep it simple for now
- [ ] An Event should be sent on startup (or when location changes) to send an Activity of Type Event with a Name of IPA.Location and Lat Long as per [this article](https://github.com/Microsoft/AI/blob/master/solutions/Virtual-Assistant/docs/virtualassistant-events.md) to the Bot. This enables the PoI skill to know "where you are" 
- [ ] Not clear how to update the configuration something else to stick in README.md

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java (hunk 2 above)
Comment (darrenj) [reply solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:72]: Locale should also be set (and configurable) - e.g. en-us and zh-cn

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java (hunk 14 above)
Comment (ryanisgrig) [reply solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/assistant_android/BotWrapper.java:60]: See https://github.com/Microsoft/AI/blob/master/solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-SpeechDevicesSDK/virtualassistant/app/src/main/java/com/microsoft/bot/solutions/speechdevices/samples/botapp/BotApplication.java for the startup events I used in the Speech DDK app.

File: solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/DirectLineChatbot.kt (hunk 9 above)
Comment (ryanisgrig) [reply solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-android/app/src/main/java/com/microsoft/directlinechatbot/DirectLineChatbot.kt:1]: I made use of the WebSocket in the OkHttp3 library: https://github.com/Microsoft/AI/blob/android-assistant/solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-SpeechDevicesSDK/virtualassistant/app/src/main/java/io/swagger/client/ServerConnection.java

**Comment by ryanisgrig** [comment]
Darren's comments cover any of my concerns. Future thoughts:

- I used the Bot Builders swagger.json to generate all of the Activity models (https://github.com/Microsoft/AI/tree/android-assistant/solutions/Virtual-Assistant/src/csharp/testharnesses/assistant-SpeechDevicesSDK/virtualassistant/app/src/main/java/io/swagger/client), we should share this.

- The DDK app uses a messaging app UI showing both bot and user utterances, but I had issues getting the adaptive cards library to build correctly.  I prefer that we continue showing the bot responses (card & text), it is easier when the speech service mispronounces words to have the text as a backup and keep demonstrations moving smoothly.

**Comment by dwhiteddsoft** [comment]
Will clean branch out

**Pull request closed without merging**